buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java

```diff
@@ -0,0 +1,302 @@
+    }
+
+    @Override
+    public void freeze() {
+        nodes.forEach(ElasticsearchNode::freeze);
+        configurationFrozen.set(true);
+    }
+
+    private void checkFrozen() {
+        if (configurationFrozen.get()) {
+            throw new IllegalStateException("Configuration for " + this + " can not be altered, already locked");
+        }
+    }
+
+    @Override
+    public void setJavaHome(File javaHome) {
+        nodes.all(each -> each.setJavaHome(javaHome));
+    }
+
+    @Override
+    public void start() {
+        String nodeNames = nodes.stream().map(ElasticsearchNode::getName).collect(Collectors.joining(","));
+        for (ElasticsearchNode node : nodes) {
+            node.defaultConfig.put("cluster.name", safeName(clusterName));
+            if (Version.fromString(node.getVersion()).getMajor() >= 7) {
+                node.defaultConfig.put("cluster.initial_master_nodes", "[" + nodeNames + "]");
+                node.defaultConfig.put("discovery.seed_providers", "file");
+            }
+        }
+        nodes.forEach(ElasticsearchNode::start);
+    }
+
+    private void writeUnicastHostsFiles() {
+        String unicastUris = nodes.stream().flatMap(node -> node.getAllTransportPortURI().stream()).collect(Collectors.joining("\n"));
+        nodes.forEach(node -> {
+            try {
+                Files.write(node.getConfigDir().resolve("unicast_hosts.txt"), unicastUris.getBytes(StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Override
+    public String getHttpSocketURI() {
+        waitForAllConditions();
+        return getFirstNode().getHttpSocketURI();
+    }
+
+    @Override
+    public String getTransportPortURI() {
+        waitForAllConditions();
+        return getFirstNode().getTransportPortURI();
+    }
+
+    @Override
+    public List<String> getAllHttpSocketURI() {
+        waitForAllConditions();
+        return nodes.stream().flatMap(each -> each.getAllHttpSocketURI().stream()).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<String> getAllTransportPortURI() {
+        waitForAllConditions();
+        return nodes.stream().flatMap(each -> each.getAllTransportPortURI().stream()).collect(Collectors.toList());
+    }
+
+    public void waitForAllConditions() {
+        long startedAt = System.currentTimeMillis();
+        logger.info("Waiting for nodes");
+        nodes.forEach(ElasticsearchNode::waitForAllConditions);
+
+        writeUnicastHostsFiles();
+
+        logger.info("Starting to wait for cluster to form");
+        addWaitForUri(
+            "cluster health yellow", "/_cluster/health?wait_for_nodes=>=" + nodes.size()  + "&wait_for_status=yellow"
+        );
+        waitForConditions(waitConditions, startedAt, CLUSTER_UP_TIMEOUT, CLUSTER_UP_TIMEOUT_UNIT, this);
+    }
+
+    @Override
+    public void stop(boolean tailLogs) {
```

Comment:
Would be nice if we could do this in such a way as we didn't have to serially block on each node. Perhaps we could pass a latch so the nodes could signal when they've stopped back to the cluster?

Is this worth it? How long does it take to gracefully shutdown a multinode cluster?

Reply:
We force kill by default since it's a throw away cluster and don't care what's on disk after it's done, so it's fast + if we do it in parallel, we'll mangle the logs of the nodes ( these are printed when a task fails )